001-clarifications-eval.txt

```diff
@@ -0,0 +1,754 @@
+    LEB128 integer that represents a byte displacement B.
+
+    The action is the same as for DW_OP_breg<N>, except that R is used
+    as the register number and B is used as the byte displacement.
+    --------------------------------------------------------------------
+
+In Section 2.5.1.2 Register Values, replace the description of
+DW_OP_regval_type with the following:
+
+    --------------------------------------------------------------------
+    DW_OP_regval_type has two operands. The first is an unsigned LEB128
+    integer that represents a register number R. The second is an
+    unsigned LEB128 integer DR that represents the byte offset of a
+    debugging information entry D relative to the beginning of the
+    current compilation unit, that provides the type T of the register
+    value.
+
+    The operation is equivalent to performing DW_OP_regx R;
+    DW_OP_deref_type DR.
+
+    > [For further discussion...]
+    > Should DWARF allow the type T to be a larger size than the size of
+    > the register R? Restricting a larger bit size avoids any issue of
+    > conversion as the, possibly truncated, bit contents of the
+    > register is simply interpreted as a value of T. If a conversion is
+    > wanted it can be done explicitly using a DW_OP_convert operation.
+    >
+    > GDB has a per register hook that allows a target specific
+    > conversion on a register by register basis. It defaults to
+    > truncation of bigger registers. Removing use of the target hook
+    > does not cause any test failures in common architectures. If the
+    > compiler for a target architecture did want some form of
+    > conversion, including a larger result type, it could always
+    > explicitly used the DW_OP_convert operation.
+    >
+    > If T is a larger type than the register size, then the default GDB
+    > register hook reads bytes from the next register (or reads out of
+    > bounds for the last register!). Removing use of the target hook
+    > does not cause any test failures in common architectures (except
+    > an illegal hand written assembly test). If a target architecture
+    > requires this behavior, these extensions allow a composite
+    > location description to be used to combine multiple registers.
+    --------------------------------------------------------------------
+
+In Section 2.5.1.3 Stack Operations, replace the description of
+DW_OP_pick with the following:
+
+    --------------------------------------------------------------------
+    DW_OP_pick has a single unsigned 1-byte operand that represents an
+    index I. A copy of the stack entry with index I is pushed onto the
+    stack.
+    --------------------------------------------------------------------
+
+Replace the description of DW_OP_over with the following:
+
+    --------------------------------------------------------------------
+    DW_OP_over pushes a copy of the entry with index 1.
```

Comment:
Did we say whether the stack starts at index 1 or at index 0?

Reply:
Yes I believe the text does define that the top stack element is index 0. But that text may be moving to 004-locations-on-stack.txt . Similar text mentioning index 1 is in the DWARF 5 standard so this makes it no worse:-)

Reply:
Our goal is to clarify, though, and merely not making it worse isn't achieving that;-)  I meanwhile found the definition in DWARF-5:
~~~
9 2.5.1.3 Stack Operations
10 The following operations manipulate the DWARF stack. Operations that index
11 the stack assume that the top of the stack (most recently added entry) has index 0.
~~~

It might make sense to move this definition to the very beginning of section 2.5.

Reply:
The concept of a numeric index into the stack only appears to be present in section "2.5.1.3 Stack Operations" so putting its definition at the top of that section seems to make sense. Moving it up to 2.5 could be done, but generally I think keeping concepts more locally defined is a good approach.